src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Parser.java

```diff
@@ -76,10 +77,47 @@
  */
 public final class Parser {
 
-    // next-line tag behaves as if it were specified on the next line
+    // Without this interface regular expressions that describe comments would have to
+    // to be standardized or needlessly complex.
 
-    private String eolMarker;
-    private Matcher markedUpLine;
+    private interface CoarseParser {
+        boolean parse(String line);
+        int payloadEnd();
+        int markupStart();
+    }
```

Comment:
I'm surprised/disappointed that this may be necessary, although I accept there is a notable difference between "whole-line comments" and "end-of-line comments".  Is it possible to reduce `CoarseParser` down to a language-specific regular expression with named groups for the payload and markup, such that the regular expression could be  a property of the `Language` object?

Reply:
I'm going to answer both of your comments, this and that one above:

> I think you can simplify the code by reducing `CoarseParser` down to a language-specific regular expression with "standard" named groups for the payload and markup.

Regular expressions with named groups were the initial design. However, I changed it to `CoarseParser` halfway through the implementation, when saw how bulky the named regex were becoming. As you also note, "end-of-line comments" and "comment lines" differ. I couldn't quickly come up with a regex that accounts for both of them.

Reply:
You may be somewhat missing the point I was trying to make.

You have two impls of `CoarseParser`, both of which contain a regular expression for the parsing, hidden inside their private matcher field.

The only other functionality of `CoarseParse` is `payloadEnd` and `markupStart`.

My suggestion is to start by updating each of the regex with named groups such that you can derive `payloadEnd` and `markupStart` from the appropriate named groups.

At that point, the only thing unique about the impls of `CoarseParser` is their regex, and that regex could become a property of the `Language` object.

> As you also note, "end-of-line comments" and "comment lines" differ. I couldn't quickly come up with a regex that accounts for both of them.

To be clear, I am _not_ suggesting a single regex.  I am suggesting a regex per supported language.

Reply:
I abstracted out the mechanics behind the `CoarseParser` precisely because I couldn't come up with a simple way to derive `payloadEnd` and `markupStart` using only groups, be they named or otherwise. My regex fu is not strong enough. If you could find a way that does not look too ugly and passes the added `TestLangProperties` test, be my guest.

Reply:
I think the following regexes with named groups should yield the same results as yours:

    "^(?<payload>.*)//(?<markup>\\s*@\\s*\\w+.+?)$"

    "^(?<payload>[ \t]*([#!].*)?)[#!](?<markup>\\s*@\\s*\\w+.+?)$"

Reply:
Above regexes do pass the test, but  I just saw (too late) that Jon posted a generic (language agnostic) solution below in the main conversation thread, which probably makes above solution obsolete.

Reply:
Your regexes are good. To me, they are conceptually simpler than those proposed by Jon. Both yours and Jon's are simpler than what I initially had.